core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+
+        let idx = handle.idx as usize;
+        unsafe {
+            slice_insert(&mut node.keys, idx, Some(key));
+            slice_insert(&mut node.vals, idx, Some(val));
+        }
+        node.len += 1;
+        node.vals[idx]
+            .as_mut()
+            .expect("value was just inserted at this position; qed")
+    }
+
+    /// Inserts a new key/value pair and an edge that will go to the right of that new pair
+    /// between this edge and the key/value pair to the right of this edge. This method assumes
+    /// that there is enough space in the node for the new pair to fit.
+    fn insert_fit_edge(&mut self, handle: KVHandle, key: K, val: V, edge: NodeHandle) {
+        self.insert_fit(handle, key, val);
+
+        let node = self
+            .get_node_mut(&handle.into())
+            .expect("node to insert (k, v, edge) into must exist");
+
+        unsafe {
+            slice_insert(&mut node.edges, handle.idx() + 1, Some(edge.node));
+        }
+
+        for idx in (handle.idx + 1)..=node.len() as u32 {
+            let handle = KVHandle::new(handle.node, idx);
+            self.correct_parent_link(handle);
+        }
+    }
+
+    fn left_child_node(&self, handle: KVHandle) -> Option<&Node<K, V>> {
+        let child = self.left_child(handle)?;
+        self.get_node(&child)
+    }
+
+    fn right_child_node(&self, handle: KVHandle) -> Option<&Node<K, V>> {
+        let child = self.right_child(handle)?;
+        self.get_node(&child)
+    }
+
+    fn left_child(&self, handle: KVHandle) -> Option<NodeHandle> {
+        let left = self.left_edge(handle);
+        self.descend(left)
+    }
+
+    fn right_child(&self, handle: KVHandle) -> Option<NodeHandle> {
+        let right = self.right_edge(handle);
+        self.descend(right)
+    }
+
+    /// Fixes the parent pointer and index in the child node below this edge. This is useful
+    /// when the ordering of edges has been changed, such as in the various `insert` methods.
+    fn correct_parent_link(&mut self, handle: KVHandle) {
+        let child = self
+            .descend(handle)
+            .expect("child in which to correct parent link must exist");
+        self.set_parent(&child, Some(handle.node), Some(handle.idx));
+    }
+
+    fn correct_childrens_parent_links(
+        &mut self,
+        handle: &NodeHandle,
+        first: usize,
+        after_last: usize,
+    ) {
+        for i in first..after_last {
+            let h = KVHandle::new(handle.node, i as u32);
+            self.correct_parent_link(h);
+        }
+    }
+
+    fn correct_all_childrens_parent_links(&mut self, handle: &NodeHandle) {
+        let node = self.get_node(handle).expect("node must exist");
+        let len = node.len();
+        self.correct_childrens_parent_links(handle, 0, len + 1);
+    }
+
+    fn has_children(&mut self, handle: &NodeHandle) -> bool {
+        let node = self.get_node(handle).expect("node must exist");
+        node.edges() > 0
+    }
```

Comment:
I know these are just internal utility and helper routines, however, I would still appreciate proper doc comments here.